### PR TITLE
feat: Adds support for On-Prem MFA Authenticator

### PR DESCRIPTION
**Pull request by gowthamidommety-okta**
## Description:

Enroll:

onPremEnroll

Verify:

onPremVerify

Passcode change:

onPremPasscodeChange

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Issue:

- [OKTA-362281](https://oktainc.atlassian.net/browse/OKTA-362281)

### CODE REVIEW

File: test/unit/spec/v2/ion/i18nTransformer_spec.js [hunk 1]
```diff
@@ -1224,4 +1225,16 @@ describe('v2/ion/i18nTransformer', function () {
       ]
     });
   });
+
+  it('gets message key', () => {
```
Comment (TianGan-okta) [inline test/unit/spec/v2/ion/i18nTransformer_spec.js:1229]: We probably also need to test challenge-authenticator and enroll-authenticator

**Review by TianGan-okta — COMMENTED** [review]
Do we also need to test uischema for the new authenticator?

File: src/v2/view-builder/views/on-prem/ChallengeAuthenticatorOnPremView.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      this.options.appState.get('currentAuthenticatorEnrollment').displayName ||
+      loc('oie.on_prem.authenticator.default.vendorName', 'login');
+    return loc('oie.on_prem.verify.title', 'login', [vendorName]);
+  },
+
+  _checkTokenChange (model, convertedErrors) {
+    const errorSummaryKeys = convertedErrors?.responseJSON?.errorSummaryKeys;
+    if (errorSummaryKeys && errorSummaryKeys.includes(ON_PREM_TOKEN_CHANGE_KEY)) {
+      // this means we are in change pin, so we should clear out answer input
+      this.model.set('credentials.passcode', '');
```
Comment (magizh-okta) [inline src/v2/view-builder/views/on-prem/ChallengeAuthenticatorOnPremView.js:27]: Why are we having to clear and rerender the whole form. We don't do this for other anyother totp  based verify or enrolll

Comment (magizh-okta) [reply src/v2/view-builder/views/on-prem/ChallengeAuthenticatorOnPremView.js:27]: Can we double check with UX/PM and remove this if they are okay.

Comment (gowthamidommety-okta) [reply src/v2/view-builder/views/on-prem/ChallengeAuthenticatorOnPremView.js:27]: Yea, i checked. This is there in v1 too. So they want parity

File: playground/mocks/data/idp/idx/authenticator-enroll-on-prem.json [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            "form": {
+              "value": [
+                {
+                  "name": "id",
+                  "required": true,
+                  "value": "autx7fdyRt87txnAs0g3",
+                  "mutable": false
+                },
+                {
+                  "name": "methodType",
```
[the remark this answers is not in the document]
Comment (gowthamidommety-okta) [reply playground/mocks/data/idp/idx/authenticator-enroll-on-prem.json:29]: yea makes sense. Will double check with them. Will send you the wiki with backend design as well

File: src/v2/ion/i18nTransformer.js [hunk 4]
```diff
@@ -100,13 +102,12 @@ const I18N_OVERRIDE_MAPPINGS = {
   'challenge-authenticator.security_question.credentials.answer': 'mfa.challenge.answer.placeholder',
   'challenge-authenticator.okta_verify.credentials.totp': 'oie.okta_verify.totp.enterCodeText',
   'challenge-authenticator.google_authenticator.credentials.otp': 'oie.google_authenticator.otp.enterCodeText',
-
+  'challenge-authenticator.del_oath.credentials.passcode': 'oie.on_prem.verify.passcode.label',
   'enroll-profile.userProfile.lastName': 'oie.user.profile.lastname',
```
Comment (TianGan-okta) [inline src/v2/ion/i18nTransformer.js:106]: nit: An empty line between `challenge-authenticator` and `enroll-profile`

File: src/v2/ion/i18nTransformer.js [hunk 5]
```diff
@@ -238,6 +239,15 @@ const getMessage = (message) => {
   return message.message;
 };
 
+/**
+ * - iff `message.i18n.key` exists return the key.
+ *
+ * @param {Message} message
+ */
+const getMessageKey = (message) => {
+  return message.i18n?.key || '';
```
Comment (TianGan-okta) [inline src/v2/ion/i18nTransformer.js:248]: Are we sure `message` will always be not null

File: test/testcafe/spec/SelectAuthenticatorForEnroll_spec.js [hunk 6]
```diff
@@ -76,6 +76,13 @@ test.requestHooks(mockEnrollAuthenticatorPassword)('should load select authentic
   await t.expect(selectFactorPage.getFactorDescriptionByIndex(4))
     .eql('Okta Verify is an authenticator app, installed on your phone or computer, used to prove your identity');
 
+  await t.expect(selectFactorPage.getFactorLabelByIndex(6)).eql('Atko Custom On-prem');
```
Comment (TianGan-okta) [inline test/testcafe/spec/SelectAuthenticatorForEnroll_spec.js:79]: Not for this PR, just a note for myself: we need to add a similar test for google authenticator

**Review by TianGan-okta — APPROVED** [review]
LGTM 🌟

File: src/v2/ion/i18nTransformer.js (hunk 4 above)
Comment (gowthamidommety-okta) [reply src/v2/ion/i18nTransformer.js:106]: sorry! so many conflicts i must've deleted this

File: src/v2/ion/i18nTransformer.js (hunk 5 above)
Comment (gowthamidommety-okta) [reply src/v2/ion/i18nTransformer.js:248]: I think so.. getMessage also assumes it. But i think i will change this. Nice catch! 👍

**Comment by haishengwu-okta** [comment]
Please double check if need to update `ViewClassNamesFactory` for className feature parity

File: src/v2/view-builder/views/on-prem/ChallengeAuthenticatorOnPremView.js (hunk 2 above)
Comment (haishengwu-okta) [reply src/v2/view-builder/views/on-prem/ChallengeAuthenticatorOnPremView.js:27]: In v1, it's checking error status code (409), are you able to do same ?

**Comment by TianGan-okta** [comment]
@haishengwu-okta @magizh-okta create a ticket for replacing the type in ViewClassNamesFactory: https://oktainc.atlassian.net/browse/OKTA-366660

**Review by magizh-okta — APPROVED** [review]

**Comment by gowthamidommety-okta** [comment]
merged as part of https://github.com/okta/okta-signin-widget/pull/1630

**Pull request closed without merging**